Aikido.Zen.Core/Models/AgentContext.cs

```diff
@@ -3,58 +3,68 @@
-        private IDictionary<string, Route> _routes = new Dictionary<string, Route>();
-        private IDictionary<string, UserExtended> _users = new Dictionary<string, UserExtended>();
-        private IDictionary<string, RateLimitingConfig> _rateLimitedRoutes = new Dictionary<string, RateLimitingConfig>();
+        private readonly ConcurrentDictionary<string, Host> _hostnames = new ConcurrentDictionary<string, Host>();
+        private readonly ConcurrentDictionary<string, Route> _routes = new ConcurrentDictionary<string, Route>();
+        private readonly ConcurrentDictionary<string, UserExtended> _users = new ConcurrentDictionary<string, UserExtended>();
+        private readonly ConcurrentDictionary<string, RateLimitingConfig> _rateLimitedRoutes = new ConcurrentDictionary<string, RateLimitingConfig>();
+        private readonly ConcurrentDictionary<string, string> _blockedUsers = new ConcurrentDictionary<string, string>();
         private Regex _blockedUserAgents;
 
-        private BlockList _blockList = new BlockList();
-        private HashSet<string> _blockedUsers = new HashSet<string>();
+        private readonly BlockList _blockList = new BlockList();
 
-        private int _requests = 0;
-        private int _attacksDetected = 0;
-        private int _attacksBlocked = 0;
-        private int _requestsAborted = 0;
-        private long _started = DateTimeHelper.UTCNowUnixMilliseconds();
+        private int _requests;
+        private int _attacksDetected;
+        private int _attacksBlocked;
+        private int _requestsAborted;
+        private readonly long _started = DateTimeHelper.UTCNowUnixMilliseconds();
         public long ConfigLastUpdated { get; set; } = 0;
         public bool ContextMiddlewareInstalled { get; set; } = false;
         public bool BlockingMiddlewareInstalled { get; set; } = false;
 
-
         public void AddRequest()
         {
-            _requests++;
+            // thread safe increment
+            Interlocked.Increment(ref _requests);
         }
 
         public void AddAbortedRequest()
         {
-            _requestsAborted++;
+            // thread safe increment
+            Interlocked.Increment(ref _requestsAborted);
         }
 
         public void AddAttackDetected()
         {
-            _attacksDetected++;
+            // thread safe increment
+            Interlocked.Increment(ref _attacksDetected);
         }
 
         public void AddAttackBlocked()
         {
-            _attacksBlocked++;
+            // thread safe increment
+            Interlocked.Increment(ref _attacksBlocked);
         }
 
         public void AddRateLimitedEndpoint(string path, RateLimitingConfig config)
         {
             if (string.IsNullOrWhiteSpace(path) || config == null)
                 return;
-            _rateLimitedRoutes[path] = config;
+            // thread safe add or update
+            _rateLimitedRoutes.AddOrUpdate(path, config, (_, __) => config);
```

Comment:
readability here is really tough, I'm not even sure if I am understanding right what is the update and what is the add